nbclient/client.py

```diff
@@ -273,7 +277,11 @@ def _kernel_manager_class_default(self):
         )
     )
 
-    def __init__(self, nb, km=None, **kw):
+    def __init__(
+            self,
+            nb: NotebookNode,
+            km: t.Optional[KernelManager] = None,
```

Comment:
Technically there's a KernelManagerABC.. but I don't *think* anyone directly inherits from the abc other than KernelManager. But it might be a good idea to change to over.

Reply:
@kevin-bates FYI I think we should shift the RemoteKernelManager to inherit from the KernelManagerABC so we can have a consistent type check / interface. From looking at the class I believe this wouldn't disrupt anything therein.

Reply:
If RemoteKernelManager inherits directly from KernelManagerABC instead of the concrete KernelManager, how does it then leverage the implementation of KernelManager? 

RMK is not an independent implementation of KernelManagerABC - otherwise that's what it would inherit from.  It needs to be plugged into the framework and that is accomplished by deriving from KernelManager.

Reply:
I tracked down the inheritance and it ends up at [MappingKernelManager](https://github.com/jupyter/notebook/blob/6e9256b0641a85baf664e846515085bac2c082a3/notebook/services/kernels/kernelmanager.py#L42)->[MultiKernelManager](https://github.com/jupyter/jupyter_client/blob/15b3b0566d962bbc042d21b858caeca2e870209a/jupyter_client/multikernelmanager.py#L43) -- which means we should instead make `MultiKernelManager` inherit the `KernelManagerABC` (OR make this line take a union of managers?) for type hinting check purposes. Drifting from the topic at hand (can move elsewhere if need be) is there a historical reason the `MappingKernelManager` is in notebook over jupyter_client? Where do we draw the line on custom extensions within the jupyter github repos for kernel managers?

> RKM is not an independent implementation of KernelManagerABC - otherwise that's what it would inherit from. It needs to be plugged into the framework and that is accomplished by deriving from KernelManager.

Think of KernelManagerABC as an interface mixin -- you can inherit from it and other things without issue using [Python's default inheritance ordering](https://data-flair.training/blogs/python-multiple-inheritance) for overlapping method names. Given the chain of existing inheritance it does reach into the jupyter_client as I pointed out above, so likely we need to just resolve that in jupyter_client without wanting/needing any changes in EG.

Reply:
Let's merge for now and we can open new issues if there's problems with EG patterns in combination.